### PR TITLE
Fix limit prefix delete case

### PR DESCRIPTION
**Pull request by cheme**
My previous PR to delete by prefix contains an ordering error, this PR fixes that.

The test change in this PR shows the problematic case.

The pr make use of the existing iter by prefix iterator for the corner case in rocksdb, and change end_prefix api to be safest, so it breaks #365.
One of the corner case that may be tricky to implement in #365 is this one, basically iterating by prefix [255] with existing key [255, 255], [255, 255, 255] ... in the db.
Ok, I am wrong this should work without change (since #365 only change read opt to set upper bound).
So this only require changing the call to end_prefix in #365

### CODE REVIEW

File: kvdb-rocksdb/src/lib.rs [hunk 1]
```diff
@@ -443,15 +443,22 @@ impl Database {
...
+								if let Some(end_range) = kvdb::end_prefix(&prefix[..]) {
+									batch.delete_range_cf(cf, &prefix[..], &end_range[..]).map_err(other_io_err)?;
+								} else {
+									for (key, _) in self.iter_from_prefix(col, &prefix[..]) {
+										batch.delete_cf(cf, &key[..]).map_err(other_io_err)?;
+									}
+								}
 							} else {
 								// Deletes all values in the column.
 								let end_range = &[u8::max_value()];
```
Comment (cheme) [inline kvdb-rocksdb/src/lib.rs:457]: Note that using [255, 255, 255... could make a smaller iteration (but would skip the test case).

File: kvdb/src/lib.rs [hunk 2]
```diff
@@ -142,15 +142,17 @@ pub trait KeyValueDB: Sync + Send + parity_util_mem::MallocSizeOf {
 
 /// For a given start prefix (inclusive), returns the correct end prefix (non-inclusive).
 /// This assumes the key bytes are ordered in lexicographical order.
-pub fn end_prefix(prefix: &[u8]) -> Vec<u8> {
+pub fn end_prefix(prefix: &[u8]) -> Option<Vec<u8>> {
```
Comment (ordian) [inline kvdb/src/lib.rs:145]: please add some docs about when `None` is returned

File: kvdb-rocksdb/src/lib.rs [hunk 3]
```diff
@@ -443,15 +443,22 @@ impl Database {
...
+						DBOp::DeletePrefix { col, prefix } => {
 							if prefix.len() > 0 {
-								let end_range = kvdb::end_prefix(&prefix[..]);
-								batch.delete_range_cf(cf, &prefix[..], &end_range[..]).map_err(other_io_err)?;
+								if let Some(end_range) = kvdb::end_prefix(&prefix[..]) {
+									batch.delete_range_cf(cf, &prefix[..], &end_range[..]).map_err(other_io_err)?;
+								} else {
+									for (key, _) in self.iter_from_prefix(col, &prefix[..]) {
+										batch.delete_cf(cf, &key[..]).map_err(other_io_err)?;
+									}
```
Comment (ordian) [inline kvdb-rocksdb/src/lib.rs:453]: I see that `delete_range` accepts only ["begin_key", "end_key"), but maybe there is a more efficient way?

Comment (cheme) [reply kvdb-rocksdb/src/lib.rs:453]: Maybe, will do some research next week.

File: kvdb-rocksdb/src/lib.rs (hunk 1 above)
Comment (ordian) [reply kvdb-rocksdb/src/lib.rs:457]: I think it's worth considering as a workaround or do you have other ideas?

Comment (cheme) [reply kvdb-rocksdb/src/lib.rs:457]: Not really, I will rework that tomorrow morning (will need to align the test case so it keeps testing the corner case).
What is the proper number of 255? 4 sounds ok

Comment (ordian) [reply kvdb-rocksdb/src/lib.rs:457]: The higher the number the better :) I was thinking more about 32 ;)

Comment (cheme) [reply kvdb-rocksdb/src/lib.rs]: I cut in the middle with 16.

File: kvdb-rocksdb/src/lib.rs [hunk 4]
```diff
@@ -443,15 +443,22 @@ impl Database {
...
 							} else {
 								// Deletes all values in the column.
-								let end_range = &[u8::max_value()];
-								batch.delete_range_cf(cf, &[][..], &end_range[..]).map_err(other_io_err)?;
-								batch.delete_cf(cf, &end_range[..]).map_err(other_io_err)?;
+								let end_range = [u8::max_value(); 16];
+								batch.delete_range_cf(cf, &prefix[..], &end_range[..]).map_err(other_io_err)?;
+								for (key, _) in self.iter_with_prefix(col, &end_range[..]) {
+									batch.delete_cf(cf, &key[..]).map_err(other_io_err)?;
+								}
```
Comment (ordian) [inline kvdb-rocksdb/src/lib.rs:461]: let's extract this into a function/closure and use it above as well, when the prefix is not empty, but is [255, 255, ...]

Comment (ordian) [reply kvdb-rocksdb/src/lib.rs:461]: although it would work for `prefix.len() < 16` case

File: kvdb-rocksdb/src/lib.rs [hunk 5]
```diff
@@ -443,15 +443,24 @@ impl Database {
...
 							batch.delete_cf(cf, &key).map_err(other_io_err)?
 						}
-						DBOp::DeletePrefix { col: _, prefix } => {
+						DBOp::DeletePrefix { col, prefix } => {
+							let iter_prefix_delete = |prefix: &[u8], batch: &mut WriteBatch| -> io::Result<()> {
+								for (key, _) in self.iter_with_prefix(col, &prefix[..]) {
+									batch.delete_cf(cf, &key[..]).map_err(other_io_err)?;
+								}
+								Ok(())
+							};
```
Comment (ordian) [inline kvdb-rocksdb/src/lib.rs:452]:
```suggestion
							let end_prefix = end_prefix(&prefix[..]);
							let end_range = end_prefix.unwrap_or_else(|| vec![u8::max_value(); 16]);
							batch.delete_cf(cf, &prefix[..], &end_range[..]).map_err(other_io_err)?;
							if end_prefix.is_none() {
								let prefix = if prefix.len() > end_range.len() { &prefix[..] } else { &end_range[..] };
								for (key, _) in self.iter_with_prefix(col, prefix) {
									batch.delete_cf(cf, &key[..]).map_err(other_io_err)?;
								}
							}
							Ok(())
```
I suggest the following code (and remove the code below). This will also optimize the case when e.g. we have several keys starting with `[255, 255, 255, 255]` and are calling `delete_prefix(&[255, 255])`, the iterator will be empty in this case, so it will issue on one range deletion.

File: kvdb-rocksdb/src/lib.rs [hunk 6]
```diff
@@ -443,15 +443,28 @@ impl Database {
...
+								}
 							} else {
 								// Deletes all values in the column.
-								let end_range = &[u8::max_value()];
-								batch.delete_range_cf(cf, &[][..], &end_range[..]).map_err(other_io_err)?;
-								batch.delete_cf(cf, &end_range[..]).map_err(other_io_err)?;
+								let end_range = [u8::max_value(); 16];
+								batch.delete_range_cf(cf, &prefix[..], &end_range[..]).map_err(other_io_err)?;
+								iter_prefix_delete(&end_range[..], &mut batch)?;
 							}
```
[the remark this answers is not in the document]
Comment (cheme) [reply kvdb-rocksdb/src/lib.rs:468]: yes, i merge in order to work locally, I am not letting it in this state :)
